lib/beaker/hypervisor/openstack.rb

```diff
@@ -250,86 +250,130 @@ def get_floating_ip
+      options = {
+        :flavor_ref => flavor(host[:flavor]).id,
+        :image_ref  => image(host[:image]).id,
+        :nics       => [{'net_id' => network(@options[:openstack_network]).id}],
+        :name       => host[:vmhostname],
+        :hostname   => host[:vmhostname],
+        :user_data  => host[:user_data] || "#cloud-config\nmanage_etc_hosts: true\n",
+        :key_name   => host[:keyname],
+      }
+      options[:security_groups] = security_groups(@options[:security_group]) unless @options[:security_group].nil?
+      vm = @compute_client.servers.create(options)
+
+      # Wait for the new instance to start up
+      try = 1
+      attempts = @options[:timeout].to_i / SLEEPWAIT
+
+      while try <= attempts
+        begin
+          vm.wait_for(5) { ready? }
+          break
+        rescue Fog::Errors::TimeoutError => e
+          if try >= attempts
+            @logger.debug "Failed to connect to new OpenStack instance #{host.name} (#{host[:vmhostname]})"
+            raise e
+          end
+          @logger.debug "Timeout connecting to instance #{host.name} (#{host[:vmhostname]}), trying again..."
+        end
+        sleep SLEEPWAIT
+        try += 1
+      end
 
-        provision_storage(host, vm) if @options[:openstack_volume_support]
-        @logger.notify "OpenStack Volume Support Disabled, can't provision volumes" if not @options[:openstack_volume_support]
+      if @options[:openstack_floating_ip]
+        # Associate a public IP to the VM
+        ip.server = vm
+        host[:ip] = ip.ip
+      else
+        # Get the first address of the VM that was just created just like in the
+        # OpenStack UI
+        host[:ip] = vm.addresses.first[1][0]["addr"]
       end
 
+      @logger.debug "OpenStack host #{host.name} (#{host[:vmhostname]}) assigned ip: #{host[:ip]}"
+    
+      # Set metadata
+      vm.metadata.update({:jenkins_build_url => @options[:jenkins_build_url].to_s,
+                          :department        => @options[:department].to_s,
+                          :project           => @options[:project].to_s })
+      @vms << vm
+
+      # Wait for the host to accept SSH logins
+      host.wait_for_port(22)
+
+      # Enable root if the user is not root
+      enable_root(host)
+
+      provision_storage(host, vm) if @options[:openstack_volume_support]
+      @logger.notify "OpenStack Volume Support Disabled, can't provision volumes" if not @options[:openstack_volume_support]
+
+    # Handle exceptions in the thread
+    rescue => e
+      @logger.error "Thread #{host} failed with error: #{e.message}"
+      # Call cleanup function to delete orphaned hosts
+      cleanup
+      # Pass the error to the main thread to terminate all threads
+      Thread.main.raise(e)
+      # Terminate the current thread (to prevent hack_etc_hosts trying to run after error raised)
+      Thread.kill(Thread.current)
+
       hack_etc_hosts @hosts, @options
```

Comment:
When does this ever run? it's in the rescue section and the thread should be killed by the line above.

Reply:
I'm testing moving the `hack_etc_hosts` to inside the main `provision` function.....

Reply:
You were correct, I sit here wrong in my wrongness...

I have moved the `hack_etc_hosts` and verified that it works as expected.  It was in fact not running at all.